_posts/2018-02-20--inception-pipelines-pt1.markdown

```diff
@@ -0,0 +1,99 @@
+
+While working on a recent project, building out a continuous deployment pipeline, I got to thinking about the number of upcoming future projects. Each project would need at least one pipeline per application, and a dedicated one for the AWS Account. Rolling a unique snowflake pipeline everytime just seemed evil and wasteful and wrong and totally against everything we stand for a [Mechanical Rock](https://www.mechanicalrock.io)!
+
+So welcome to the Inception Pipeline; a [CloudFormation](https://aws.amazon.com/cloudformation/) template that plants itself inside an AWS Account and then self manages and self updates itself using nothing more than off-the-shelf AWS services.
+
+## What technologies are we going to use
+
+* [CloudFormation](https://aws.amazon.com/cloudformation/)
+* [CodeCommit](https://aws.amazon.com/codecommit/)
+* [CodePipeline](https://aws.amazon.com/codepipeline/)
+* [IAM](https://aws.amazon.com/iam/) [Roles](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles.html)
+* [S3](https://aws.amazon.com/s3/)
+
+## What are the prerequisites
+
+1. An AWS Account to create the pipeline in.
+1. The AWS CLI installed and configured with [access credentials](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html).
+1. CodeCommit [git credentials configured](https://docs.aws.amazon.com/codecommit/latest/userguide/setting-up-https-unixes.html#setting-up-https-unixes-credential-helper).
+
+While not strictly required, a passing familarity of Bash Shell scripts, Git, JSON & YAML, and CloudFormation templates will make understanding everything easier.
+
+## How it all works
+
+At a high-level, the Inception Pipeline works by executing a CloudFormation template which then creates a CodeCommit repository, a CodePipeline pipeline and a few other supporting resources. The first non-source action in the pipeline is a [CloudFormation Deployment Action](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/continuous-delivery-codepipeline.html). This blog post won't dive deeply into the CloudFormation template (I'll leave that as an exercise for you dear reader). Instead I'll just discuss the really juicy bits.
+
+The secret-sauce to the Inception Pipeline is the using the same CloudFormation stackname in the CLI/console invocation and the CodePipeline Action (CloudFormation parameter ```StageAdministerPipelineStackName```). Once the initial CloudFormation stack has been created, the CodePipeline action below maintains the pipeline.
+
+```yaml
+- Name: 'AdministerPipeline'
+    Actions:
+    - Name: 'AdministerPipeline'
+        ActionTypeId:
+            Category: 'Deploy'
+            Owner: 'AWS'
+            Provider: 'CloudFormation'
+            Version: '1'
+        Configuration:
+            ActionMode: 'REPLACE_ON_FAILURE' 
+            Capabilities: 'CAPABILITY_NAMED_IAM'
+            RoleArn: !GetAtt [CloudFormationDeployActionRole, Arn]
+            StackName: !Ref StageAdministerPipelineStackName
```

Comment:
Add a comment above here to draw the readers attention:

`# SOMETHING_SHOUTY - blah...`